styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java

```diff
@@ -89,17 +80,26 @@ public void transitionInto(RunState state) {
             } else {
               LOG.error(msg, e);
             }
-            stateManager.receive(Event.runError(state.workflowInstance(), e.getMessage()), 
-                state.counter() + 1);
+            stateManager.receive(Event.runError(state.workflowInstance(), e.getMessage()), state.counter());
           } catch (IsClosedException isClosedException) {
             LOG.warn("Failed to send 'runError' event", isClosedException);
           }
+          return;
+        }
+
+        // Emit `submitted` _after_ starting execution to ensure that we retry in case of failure.
+        final Event submitted = Event.submitted(state.workflowInstance(), runSpec.executionId());
+        try {
+          stateManager.receive(submitted, state.counter());
+        } catch (IsClosedException isClosedException) {
+          LOG.warn("Could not emit 'submitted' event", isClosedException);
         }
         break;
 
       case TERMINATED:
       case FAILED:
       case ERROR:
+        // TODO: remove this effectively unused cleanup?
```

Comment:
We should remove `cleanup` in DockerRunner interface as well if we remove this. It was kept for separation of concern.

Reply:
I'd be happy to remove that and the rotting/unused `LocalDockerRunner` as well and just require k8s. For local execution there's minikube and docker desktop (with k8s bundled) anyway.

Reply:
I can leave it to another PR.

Reply:
Yeah